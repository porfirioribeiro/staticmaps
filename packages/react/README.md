### Render static maps with React

This small package allows you to render static maps easly with react.

It uses svg to render the tiles and the vectors for the Polygon. You can specify a bounding box or if you pass any shape, it will automatically find the bounds for the shapes you want to add.

![demo](demo.png)

## Instalation

> yarn add @staticmaps/react

## Usage


```typescript
<StaticMap
  width={250}
  height={250}
  padding={[5, 5]}
  tileProvider={esriWorldImageryTileProvider}
  polygons={[rosaMP]}
  markers={[rosaMarker1]}
/>
```

## Available shapes

- Base TileLayer using x,y,z
- Marker
- Polygon
- OverlayImage

## Demo

See a simple demo [here](https://staticmaps-react.netlify.app/)

> pnpm start

