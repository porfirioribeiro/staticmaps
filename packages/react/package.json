{
  "name": "@staticmaps/react",
  "version": "1.0.0",
  "author": "Porf√≠rio Ribeiro <porfirioribeiro@gmail.com>",
  "license": "MIT",
  "type": "module",
  "source": "src/index.ts",
  "exports": {
    "require": "./dist/index.cjs.js",
    "import": "./dist/index.es.js"
  },
  "main": "./dist/index.cjs.js",
  "module": "./dist/index.es.js",
  "unpkg": "./dist/index.iife.js",
  "esmodule": "dist/index.es.js",
  "umd:main": "dist/index.umd.js",
  "types": "dist/index.d.ts",
  "dependencies": {
    "@staticmaps/core": "^1.0.0",
    "react": "^17.0.2",
    "react-dom": "^17.0.2"
  },
  "peerDependencies": {
    "@staticmaps/core": "^1.0.0",
    "react": "^17.0.1"
  },
  "scripts": {
    "start": "vite serve ./demo/",
    "build": "vite build",
    "test": "vitest run  --coverage",
    "format": "prettier src --write"
  },
  "devDependencies": {
    "@testing-library/jest-dom": "^5.16.2",
    "@testing-library/react": "^12.1.4",
    "@types/node": "^17.0.22",
    "@types/react": "^17.0.41",
    "@types/react-dom": "^17.0.14",
    "@vitejs/plugin-react": "^1.2.0",
    "c8": "^7.11.0",
    "jsdom": "^19.0.0",
    "prettier": "^2.6.0",
    "typescript": "^4.6.2",
    "vite": "^2.8.6",
    "vite-plugin-dts": "^0.9.10",
    "vitest": "^0.7.7"
  }
}
