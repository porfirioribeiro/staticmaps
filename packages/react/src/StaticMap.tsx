import React, { useReducer } from 'react';
import { StaticMapsState, StaticMapOptions, createStaticMap } from '@staticmaps/core';

function reducer(state: StaticMapsState, _action: any): StaticMapsState {
  return state;
}

interface StaticMapProps extends StaticMapOptions {
  className?: string;
  style?: React.CSSProperties;
}

export function StaticMap(props: StaticMapProps) {
  const { className, style } = props;
  delete props.className;
  delete props.style;
  const [ctx] = useReducer(reducer, props, createStaticMap);

  const { width, height, viewBox, tiles, multiPolygons, overlayImages, markers } = ctx;

  return (
    <svg width={width} height={height} viewBox={viewBox} className={className} style={style}>
      {tiles.map((t, key) => React.createElement('image', Object.assign({ key }, t)))}
      {overlayImages.map((oi, key) => React.createElement('image', Object.assign({ key }, oi)))}
      {multiPolygons.map((mp, key) => React.createElement('path', Object.assign({ key }, mp, { vectorEffect: 'non-scaling-stroke' })))}
      {markers.map((m, key) => React.createElement('image', Object.assign({ key }, m)))}
    </svg>
  );
}
