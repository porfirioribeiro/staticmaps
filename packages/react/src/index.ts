export * from '@staticmaps/core';
export { StaticMap } from './StaticMap';
