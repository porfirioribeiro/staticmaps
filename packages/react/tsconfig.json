{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "jsx": "preserve",
    "moduleResolution": "Node"
  },
  "include": ["src"]
}
