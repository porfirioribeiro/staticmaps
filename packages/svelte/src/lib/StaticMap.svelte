<script lang="ts">
	import { createStaticMap, type StaticMapOptions, type StaticMapsState } from '@staticmaps/core';

	export let options: StaticMapOptions;

	let { width, height, viewBox, tiles, multiPolygons } = {} as StaticMapsState;

	$: {
		({ width, height, viewBox, tiles, multiPolygons } = createStaticMap(options));
		console.log('updates', width, height);
	}
</script>

<svg {width} {height} {viewBox}>
	{#each tiles as tile}
		<image {...tile} />
	{/each}
	{#each multiPolygons as multiPolygon}
		<path {...multiPolygon} vector-effect="non-scaling-stroke" />
	{/each}
</svg>
