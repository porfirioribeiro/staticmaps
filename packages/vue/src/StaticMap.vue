<script setup lang="ts">
import { StaticMapsState, StaticMapOptions, createStaticMap } from '@staticmaps/core';

const props = defineProps<{ options: StaticMapOptions }>();
const o = createStaticMap(props.options);
</script>

<template>
  <svg :width="o.width" :height="o.height" :viewBox="o.viewBox">
    <image v-for="tile in o.tiles" v-bind="tile"></image>
    <path v-for="multiPolygon in o.multiPolygons" v-bind="multiPolygon" vectorEffect="non-scaling-stroke"></path>
  </svg>
  <div>Hello</div>
</template>
