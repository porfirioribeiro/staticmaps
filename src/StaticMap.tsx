import React, { useReducer } from "react";
import { StaticMapsState, StaticMapProps } from "./types";
import { createStaticMap } from "./createStaticMap";

function reducer(state: StaticMapsState, action: any): StaticMapsState {
  return state;
}

export function StaticMap(props: StaticMapProps) {
  const [ctx] = useReducer(reducer, props, createStaticMap);

  const {
    width,
    height,
    viewBox,
    tiles,
    multiPolygons,
    overlayImages,
    markers,
  } = ctx;

  return (
    <svg width={width} height={height} viewBox={viewBox}>
      {tiles.map((t, i) => (
        <image key={i} {...t} />
      ))}
      {overlayImages.map((oi, i) => (
        <image key={i} {...oi} />
      ))}
      {multiPolygons.map((mp, i) => (
        <path key={i} {...mp} vectorEffect="non-scaling-stroke" />
      ))}
      {markers.map((m, i) => (
        <image key={i} {...m} />
      ))}
    </svg>
  );
}
