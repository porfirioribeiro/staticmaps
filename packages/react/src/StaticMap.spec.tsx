import React from 'react';
import { render } from '@testing-library/react';

import '@testing-library/jest-dom';
import { StaticMap } from './StaticMap';

test('renders', () => {
  const { container } = render(
    <StaticMap
      width={345}
      height={138}
      padding={[10, 10]}
      bbox={[-61.0698239073564793, -21.3991246364298995, -61.0529563571889184, -21.407334765312239]}
    />,
  );
  expect(container).toMatchSnapshot();
});
