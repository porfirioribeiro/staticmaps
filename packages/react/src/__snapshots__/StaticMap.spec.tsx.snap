// Vitest Snapshot v1

exports[`renders 1`] = `
<div>
  <svg
    height="138"
    viewBox="0 0 345 138"
    width="345"
  >
    <title>
      &copy; &lt;a href="https://www.openstreetmap.org/copyright"&gt;OpenStreetMap&lt;/a&gt; contributors
    </title>
    <image
      height="256"
      href="https://tile.openstreetmap.org/14/5412/9189.png"
      width="256"
      x="-91"
      y="-75"
    />
    <image
      height="256"
      href="https://tile.openstreetmap.org/14/5413/9189.png"
      width="256"
      x="165"
      y="-75"
    />
  </svg>
</div>
`;
